x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedConfig.java

```diff
@@ -552,9 +592,13 @@ private static ChunkingConfig defaultChunkingConfig(@Nullable AggProvider aggPro
         if (aggProvider == null || aggProvider.getParsedAggs() == null) {
             return ChunkingConfig.newAuto();
         } else {
-            long histogramIntervalMillis = ExtractorUtils.getHistogramIntervalMillis(aggProvider.getParsedAggs());
+            AggregationBuilder histogram = ExtractorUtils.getHistogramAggregation(aggProvider.getParsedAggs().getAggregatorFactories());
+            if (histogram instanceof CompositeAggregationBuilder) {
+                return ChunkingConfig.newOff();
```

Comment:
OK, after doing some experimentation, ON an index with `14479391` docs, and grouping by 15m intervals + on a term field with ~851502, chunking provides a much better experience. 

I am thinking there should be an "auto", similar to what we do with scroll that takes into account the number of group_by fields + the date_histogram interval, min max time, and possibly the cardinality of the terms.

Reply:
Oh yeah, chunking by 1000 * <date_histogram_interval> for this job made it run like 3 times faster than no chunking. 

`12m` vs `44m`

But still twice as slow as regular scroll (`6m`).

Reply:
OK, I think I hit a sweet spot. 

Let `x` be the configured cardinality bucket size.
Let `t` be the date_histogram time interval
Let `[y]` be the cardinalites of the term group bys

`(x * t)/(sum([y])/x)` The `sum` is where I am not sure. We want to capture things with HUGE cardinality as this causes a GIANT number of docs to be grabbed and sorted. So, reducing the time span really helps with that as there is smaller work. I feel like doing a `product` would result is reducing the chunking size by TOO much. But, then again, if you are grouping by two terms, each with ~800k+ cardinality, I am not sure. That also assumes that these time buckets have uniform cardinality, which they don't. So, maybe a sum or an average would be good enough.

So, in the above example, 15 minute interval with a single term group by with a `~851502` cardinality
`~1057579ms` timespan. 
(1000 * 15m) = 900,000,000ms
(851502/1000) = 851
inter


This job finished in ~`3m`. 

@dimitris-athanasiou @tveasey ^ let me know what you think.

Reply:
The total number of distinct terms is somewhere between `max(y)` and `min(# docs, prod(y))`.  For definite product would be at the wrong extreme and should be truncated to # docs. This is a pretty large range and `sum([y])` is probably a better choice all assumes strong correlation between categories. Also note that any case you have several high cardinality fields `min(# docs, prod(y)) = # docs`.

So my thoughts...

First of all would it be prohibitive to actually just distinct count concatenated group by fields? You could actually grab a few different time buckets and just use the mean of the distinct counts of tuples on these time buckets. This means you wouldn't have to search the entire range to get this. You could even use this quantity to decide if you should just be scrolling the data, i.e. if average dc -> average doc count. 

The expected number of distinct tuples assuming things appear independently at random is slightly fiddly combinatorics (it can be worked out from the answer to [this](https://math.stackexchange.com/questions/72223/finding-expected-number-of-distinct-values-selected-from-a-set-of-integers) problem using, following your notation, `n = prod([y])` and assuming you draw m = # docs times). But the assumption that things appear totally independently a random is probably too pessimistic. However, once you depart from this assumption what correlation does it make sense to choose? I'm not sure, which is why if we can actually get the correct distinct counts as above I think this is preferable. Sum assumes rather strong correlation between individual values. Something which potentially makes a less strong assumption would be he following:

let `y_(i)` are the distinct counts ordered by descending cardinality then 

`E[distinct tuples]= exp(sum_{i in [0, n-1]}{ 1 / k^i log(y_(i)) }) with k >= 1`

If k is large this won't be too much bigger than `max(y)` if it is near 1 it'll tend to `prod(y)`. Again one should always impose the hard constraint `E[distinct tuples] < # docs`.

Reply:
See: https://github.com/elastic/elasticsearch/issues/70035

Reply:
I think the logic will have to be (until #70035 gives us more options):
- Using no chunking if the datafeed has no query (match_all)
- Using some chunking heuristic if the datafeed has a query

Reply:
Doing some more tests, It seems that `composite` with just a `date_histogram` (when using a filter query) is slower than using a simple `date_histogram`. I am thinking this has to do with `composite` needing to know latest bucket and consequently sorting the information.

Consequently our normal heuristic of 1000 buckets chunks cannot really be the starting point for composite aggs. 

Still digging around here to see what is truly good enough for now. 

@dimitris-athanasiou @tveasey any more thoughts here?

Reply:
I am thinking that all this worrying is getting much too complicated. 

The real solution here is to optimize the composite agg implementation, not the datafeed chunking. 

I am thinking this piecewise calculation (`y` is the set of term field cardinalities, `page_size` is the composite agg paging size).

- If the query has no filters (just a match_all), we use the NO chunking as composite aggs is optimized (should get more lax in time)
- If `(prod(y)) > page_size`; chunking size is the bucket_span of the job. 
- If  `(prod(y)) <= page_size`; page_size * date_histogram bucketing size.

The two concerns here:

- If page_size is only `10` (the default value), we do many more searches
- If `prod(y)` is equal to an exceptionally large page size AND the date_histogram interval is << bucket_span, our searches will be WAY too slow.

But, I think this is a "good enough" fix. We need to get composite aggs to be optimized for a real way forward here.

Reply:
I agree this complexity is not something that belongs in the extractor's side of things. I suggest we move forward supporting composite aggs without doing any optimization. By default, chunking should be `off` when composite aggs are provided.

Any effort to optimize this now creates complicated test scenarios that should be removed when the composite agg itself is improved. Unlike normal scrolling, which was crucial for the typical ML use case, I think supporting composite aggs provides an optional alternative for advanced users. As such, we can refrain from adopting this complexity at the moment.